lib/provider/awssamlprovider.go

```diff
@@ -0,0 +1,472 @@
+	// It will be ignored.
+	SessionCacheSingleItem bool
+}
+
+// validates aws saml configuration options.
+func (o *AWSSAMLProviderOptions) Validate() error {
+	if o.SessionDuration < MinSessionDuration {
+		return errors.New("Minimum session duration is " + MinSessionDuration.String())
+	} else if o.SessionDuration > MaxSessionDuration {
+		return errors.New("Maximum session duration is " + MaxSessionDuration.String())
+	}
+	if o.AssumeRoleDuration < MinAssumeRoleDuration {
+		return errors.New("Minimum duration for assumed roles is " + MinAssumeRoleDuration.String())
+	} else if o.AssumeRoleDuration > MaxAssumeRoleDuration {
+		log.Println(o.AssumeRoleDuration)
+		return errors.New("Maximum duration for assumed roles is " + MaxAssumeRoleDuration.String())
+	}
+
+	return nil
+}
+
+// updates aws saml configuration with package provided defaults.
+func (o *AWSSAMLProviderOptions) ApplyDefaults() {
+	if o.AssumeRoleDuration == 0 {
+		o.AssumeRoleDuration = DefaultAssumeRoleDuration
+	}
+	if o.SessionDuration == 0 {
+		o.SessionDuration = DefaultSessionDuration
+	}
+}
+
+// creates a new AWS saml provider
+func NewAWSSAMLProvider(sessions SessionCacheInterface, profile string, opts AWSSAMLProviderOptions, oktaClient OktaClient, selector SAMLRoleSelection) (*AWSSAMLProvider, error) {
+	var profileARN string
+	var err error
+
+	opts.ApplyDefaults()
+	if err := opts.Validate(); err != nil {
+		return nil, err
+	}
+
+	log.Debug("Provider Option is deprecated: SessionCacheSingleItem")
+
+	source := lib.SourceProfile(profile, opts.Profiles)
+
+	// if the assumable role is passed it have it override what is in the profile
+	if opts.AssumeRoleArn != "" {
+		profileARN = opts.AssumeRoleArn
+		log.Debug("Overriding Assumable role with: ", profileARN)
+	} else {
+		// if `role_arn` isn't provided as part of the profile we can still prompt
+		// for it later after we get the saml assertion and know all the roles the
+		// user can assume.
+		profileARN = opts.Profiles[source]["role_arn"]
+	}
+
+	provider := AWSSAMLProvider{
+		AWSSAMLProviderOptions: opts,
+		oktaClient:             oktaClient,
+		profileARN:             profileARN,
+		sessions:               sessions,
+		profile:                profile,
+		selector:               selector,
+	}
+
+	if region := opts.Profiles[source]["region"]; region != "" {
+		provider.awsRegion = region
+	}
+	err = provider.getSAMLURL()
+	if err != nil {
+		return nil, err
+	}
+	return &provider, nil
+}
+
+// Gets a set of STS credentials to access AWS services.
+func (p *AWSSAMLProvider) Retrieve() (credentials.Value, error) {
+	window := p.ExpiryWindow
+	if window == 0 {
+		window = time.Minute * 5
+	}
+
+	// TODO(nick): why are we using the source profile name and not the actual profile's name?
+	source := lib.SourceProfile(p.profile, p.Profiles)
+	profileConf, ok := p.Profiles[p.profile]
+	if !ok {
+		return credentials.Value{}, fmt.Errorf("missing profile named %s", p.profile)
+	}
+	key := sessioncache.OrigKey{
```

Comment:
I believe this will get the wrong session since it doesn't take the role override into account. https://github.com/segmentio/aws-okta/pull/237